add notice about install prerequisites

dmitris opened the pull request:
add notice about installing rpmbuild with rpm-build package as 
prerequisite for building RPMs.

<!-- The following line must be included in your pull request -->
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

qbarnes commented:
I don't see what you're going to the effort of listing only one of numerous packages necessary for building the RPMs natively.

If you want to have an explicit list, then I'd suggest making the list complete.  And before going down the rabbit hole of listing dozens of packages, may I suggest just installing the yum groups "Development Tools" and "RPM Development Tools" to take care of virtually everything including `rpm-build`.

But of course, my above suggestions only work for RHEL and RHEL-based operating systems.  If you're going to cover details on building RPMs, you should also cover other RPM-based operating systems like openSUSE too.

dmitris commented:
thanks @qbarnes - I updated the PR and added yum groups.  I don't see the "RPM Development Tools" group on rhel7, only on rhel8, therefore I split the instructions in two parts for rhel7 and rhel8.

I agree it would be great to have installation instructions for multiple OSes but I think it would be better to have them at least for RHEL to start then nothing.  Hopefully the interested openSUSE users could contribute additions for that OS 😄 

Eventually I'd like to be able to use the `INSTALL` file to take a "clean slate" server or VM and get to the state where you can build the modules and packages by copying the given commands. I may be missing some dependencies like `iptables-devel`, will double-check in the next days.

qbarnes commented:
> I don't see the "RPM Development Tools" group on rhel7, only on rhel8, therefore I split the instructions in two parts for rhel7 and rhel8.

That's because on RHEL/CentOS 7 the `rpm-build` package is part of the "Development Tools" package group.

> Eventually I'd like to be able to use the INSTALL file to take a "clean slate" server or VM and get to the state where you can build the modules and packages by copying the given commands.

You're trying to recreate and duplicate exactly what `mock` already does automatically for you.  Use `mock` instead!

Anyway, telling people to install `rpm-build` to build RPMs is like telling people they need to install `gcc` to build a C program.  If they're building RPMs, that's very basic knowledge they should already have.

> I may be missing some dependencies like iptables-devel, will double-check in the next days.

No, you don't need to specify those because they're in BuildRequires rules of the spec file.  They'll automatically be installed as part of the build.  You'd only need to manually install them if you're doing non-package builds.

I suggest before submitting any more pull requests for building that you study the basics on how general RPM builds work first.  Red Hat has a class on this.  Check out "From Source to RPM in 120 Minutes" here: https://gitlab.com/redhatsummitlabs/from-source-to-rpm-in-120-minutes and here: https://rpm-packaging-guide.github.io/

dmitris commented:
> You'd only need to manually install them if you're doing non-package builds.

wouldn't that be a valid use case as well?  Imagine someone comes across this repo who wants to know how to get from the fresh git checkout to the built software and running tests - I believe it would be useful and more "user-friendly" to have some instructions that specify the prerequisites and the steps to build and run tests, even though one can "extract" them from the requirements in the spec files (or from the build errors etc.).

I can see your point about `rpmbuild` notice being "too elementary" 😄  Rephrased to follow your previous suggestion (plus added a pointer to the spec files regarding the dependencies):
> may I suggest just installing the yum groups "Development Tools" and "RPM Development Tools" to take care of virtually everything including rpm-build.

Do you think it would be worth to have some notice of "prerequisites" such as iptables devel package and kernel headers on the "top level", for example in the INSTALL file at the repo root?

qbarnes commented:
> wouldn't that be a valid use case as well? Imagine someone comes across this repo who wants to know how to get from the fresh git checkout to the built software and running tests - I believe it would be useful and more "user-friendly" to have some instructions that specify the prerequisites and the steps to build and run tests, even though one can "extract" them from the requirements in the spec files (or from the build errors etc.).

No, you don't have to extract anything from a `.spec` file.  Again, you're apparently missing a basic understanding of RPMs and building.  There's already a tool that does it for you, `yum-builddep`.  When you give it a `.spec` file like is provided in this repo, it installs for you all the RPMs you need for building.  A version of that tool is how `mock` correctly populates the build chroot for you.

Anytime you think you need to invent a solution for a what should be a common dev problem, the first question to yourself should be, "Has anyone already cleanly solved this?"

Along with the "120 minutes" class, I would suggest you run `man -k yum` and `man -k rpm` and familiarize yourself with the commands listed.  I still run those two commands from time to time to refresh my knowledge, and occasionally, find a new command to go off and tinker with.

Anytime I'm trying to build some random FOSS package I just downloaded the code for and there's no build instructions at all, it's not even an inconvenience.  I don't have to try to determine dependency package lists.  If the build fails because of a missing command, I just type `yum install /usr/bin/<cmd>`.  If the compile fails due to a missing header file, I just type `yum install /usr/include/<hdrpath>`.  Same works for a missing library, just give `yum` the path and it'll figure out the needed packages.

Even when package lists are included in build instructions, they're often outdated or not applicable to the distro and version I'm trying to build for, so they mostly get ignored by me anyway.  The above technique is often easier and more reliable.

But always keep in mind my one rule I tell devs when building software for RPMs (especially in a professional environment), "If you're not using `mock` to build your RPMs, you're doing it wrong."  Devs should be encouraged to use `mock` whenever possible, not doing builds natively on a system.

> Do you think it would be worth to have some notice of "prerequisites" such as iptables devel package and kernel headers on the "top level", for example in the INSTALL file at the repo root

No.  If someone is wanting the build and use `iptables-daddr` and doesn't have basic build knowledge of knowing how to determine the necessary packages for building for their particular distro and release, they probably shouldn't be trying to build and deploy L3DSR.  It's like having a friend and you want to go for a drive in your car and they offer to do the driving.  You hand them the keys, then while sitting in the driver's seat, complain that you're not telling them how to start the car and put it in drive.  What would you do?  I would take my keys back and advise them to take a Driver's Ed class.  If they don't know the basics of how to start the car and put it in gear, they're going to be in for a lot more trouble once the car is in motion.

In the nearly a decade that `iptables-daddr` has been FOSS'd here, I've fielded many dozens of random questions from users of the software.  Some of them were build questions, but out of all of people I've helped, they weren't looking for such basic build knowledge as to what dependent packages they needed to install.  L3DSR is not some GUI text editor or GNOME widget that ordinary, everyday users might want to build for their own use.  Deploying L3DSR requires some solid, professional admin and dev  skills.  If they don't know enough to compile it without hand-holding, they're going to be missing a lot of other basic knowledge that'll trip them up when trying to deploy it to a production environment.

dmitris commented:
>I'm trying to build some random FOSS package I just downloaded the code for and there's no build instructions at all, it's not even an inconvenience. 

I probably have a different "mindset" - I do appreciate and find it helpful when FOSS package owners provide reasonably detailed and complete installation and/or build instructions.  Let's put ourselves in the shoes or someone who just learned about this repository and is looking at https://github.com/yahoo/l3dsr for the first time - the README page explains how [L3]DSR works and who wrote it (which is great) but doesn't contain:
* any build instructions or pointers where to look for them
* an overview of the project structure, for example to tell the reader to look at https://github.com/yahoo/l3dsr/blob/master/linux/dsrtools/README and https://github.com/yahoo/l3dsr/blob/master/linux/iptables-daddr/README and the corresponding `INSTALL`/`USING` files
* there is no CI setup such as `.travis.yml` file or GitHub Actions config which would "codify" the build, test, and possibly package steps and could also be used as reference

I tried my best to consider your arguments carefully and "in a good faith" but still believe that adding a few "extra documentation" details like above would help to make the project more "user-friendly" for new users or potential contributors. But let's close this now and not waste any more time on it! 😄

qbarnes commented:
> the README page explains how [L3]DSR works and who wrote it (which is great) but doesn't contain ...

A good point.  I didn't write the original `README` and it's showing its age.  If you'd like to open an issue or submit a pull request with the relevant information on the overall structure of the repo, that might be good.  Maybe create a top-level `CONTENTS.md` file and reference it in the `README` (and maybe convert the `README` to a `README.md` while at it).

I think there are a lot of ways documentation in this repo could be improved, but the 2 line patch of this PR I felt wasn't a worthwhile effort.

On a different GitHub FOSS project I own I tried setting up Travis CI.   That project was somewhat straight-forward Linux C code utility.  I wanted Travis to just do a simple build for me off `master` and publish it as a new release.  After a few days of bashing my head, I gave up.  Though native binary C builds with Travis are supposed to be supported, I quickly realized Travis wasn't designed around doing those builds and had a rather long list of shortcomings.  But a lot of my frustration also came from how absolutely horribly slow the build turn around time was on the free version of Travis and how completely unhelpful the limited, confusing error messages were.  If you think you can write a good, helpful `.travis.yml` file that would work for this repo's code without going insane, please give it a try.

You may not know, but the majority of users of this software/repo are on Ubuntu.  Ubuntu doesn't have the (relatively) stable kABI of RHEL/CentOS.  I would have liked to put together and support a DKMS build that would be usable on most any distro, but that's just another effort I was suddenly forced to drop on the floor.  Such a build target would automate several more complex build steps for such users.  And I'm sure such users would also love a `.deb` file.  That's part of the reason I didn't bother with any detailed build documentation.  Much of it was meant to be captured in code by a `.deb` file and doing DKMS builds.  Then document those higher level build steps largely in part by offering Ubuntu/Fedora/CentOS/openSUSE example build jobs running on Open Build Service (that was a time before GitHub offered any build service).

qbarnes commented:
BTW, you may find my literal "Hello, World" RPM walkthrough using `rpmbuild` and `mock` informative: https://github.com/qbarnes/hello-rpm
